From 19564e696aca506a14ceb4a08aa2b7c7988a2473 Mon Sep 17 00:00:00 2001
From: Arnold D. Robbins <arnold@skeeve.com>
Date: Fri, 25 Apr 2014 10:21:43 +0000
Subject: In forked children, set SIGPIPE to SIG_DFL.

---
* io.c (two_way_open): In forked child, reset SIGPIPE to SIG_DFL.
Fixes problems with "broken pipe" errors from child processes,
restoring 4.1.0 and earlier behavior. Thanks to Daryl F
<wyatt@prairieturtle.ca> for the report.
(gawk_popen): Ditto.

diff --git a/io.c b/io.c
index a6e786f..2bb8f28 100644
--- a/io.c
+++ b/io.c
@@ -2078,6 +2078,7 @@ use_pipes:
 		    || close(ctop[0]) == -1 || close(ctop[1]) == -1)
 			fatal(_("close of pipe failed (%s)"), strerror(errno));
 		/* stderr does NOT get dup'ed onto child's stdout */
+		signal(SIGPIPE, SIG_DFL);
 		execl("/bin/sh", "sh", "-c", str, NULL);
 		_exit(errno == ENOENT ? 127 : 126);
 	}
@@ -2271,6 +2272,7 @@ gawk_popen(const char *cmd, struct redirect *rp)
 			fatal(_("moving pipe to stdout in child failed (dup: %s)"), strerror(errno));
 		if (close(p[0]) == -1 || close(p[1]) == -1)
 			fatal(_("close of pipe failed (%s)"), strerror(errno));
+		signal(SIGPIPE, SIG_DFL);
 		execl("/bin/sh", "sh", "-c", cmd, NULL);
 		_exit(errno == ENOENT ? 127 : 126);
 	}
--
cgit v0.9.0.2
